src/java.base/share/classes/jdk/internal/misc/MainMethodFinder.java

```diff
@@ -0,0 +1,150 @@
+
+        return argc == 0 || argc == 1 && method.getParameterTypes()[0] == String[].class;
+    }
+
+    /**
+     * Gather all the "main" methods in the class hierarchy.
+     *
+     * @param declc  the top level declaring class
+     * @param refc   the declaring class or super class
+     * @param mains  accumulated main methods
+     */
+    private static void gatherMains(Class<?> declc, Class<?> refc, List<Method> mains) {
+        if (refc != null && refc != Object.class) {
+            for (Method method : refc.getDeclaredMethods()) {
+                // Must be named "main", public|protected|package-private, not synthetic (bridge) and either
+                // no arguments or one string array argument.
+                if ("main".equals(method.getName()) &&
+                        !method.isSynthetic() &&
+                        !Modifier.isPrivate(method.getModifiers()) &&
+                        correctArgs(method))
+                {
+                    mains.add(method);
+                }
+            }
+
+            gatherMains(declc, refc.getSuperclass(), mains);
+        }
+    }
+
+    /**
+     * Comparator for two methods.
+     * Priority order is;
+     * sub-class < super-class.
+     * static < non-static,
+     * public < non-public,
+     * string arg < no arg and
+     *
+     * @param a  first method
+     * @param b  second method
+     *
+     * @return -1, 0 or 1 to represent higher priority. equals priority or lesser priority.
+     */
+    private static int compareMethods(Method a, Method b) {
+        Class<?> aClass = a.getDeclaringClass();
+        Class<?> bClass = b.getDeclaringClass();
+
+        if (aClass != bClass) {
+            if (bClass.isAssignableFrom(aClass)) {
+                return -1;
+            } else {
+                return 1;
+            }
+        }
+
+        int aMods = a.getModifiers();
+        int bMods = b.getModifiers();
+        boolean aIsStatic = Modifier.isStatic(aMods);
+        boolean bIsStatic = Modifier.isStatic(bMods);
+
+        if (aIsStatic && !bIsStatic) {
+            return -1;
+        } else if (!aIsStatic && bIsStatic) {
+            return 1;
+        }
+
+        boolean aIsPublic = Modifier.isPublic(aMods);
+        boolean bIsPublic = Modifier.isPublic(bMods);
+
+        if (aIsPublic && !bIsPublic) {
+            return -1;
+        } else if (bIsPublic && !aIsPublic) {
+            return 1;
+        }
+
+        int aCount = a.getParameterCount();
+        int bCount = b.getParameterCount();
+
+        if (bCount < aCount) {
+            return -1;
+        } else if (aCount < bCount) {
+            return 1;
+        }
+
+        return 0;
+    }
+
+    /**
+     * {@return priority main method if none found}
+     *
+     * @param mainClass main class
+     *
+     * @throws NoSuchMethodException when not preview and no method found
+     */
+    public static Method findMainMethod(Class<?> mainClass) throws NoSuchMethodException {
```

Comment:
The latest iteration of changes removes the warning but the latest iteration of the JEP proposes a warning when an instance main or static no-args main is preferred over an inherited static main. Is it that the JEP is just running ahead a bit and the changes in the PR will catch up soon?

Reply:
Yes. There is some juggling going on to get the code to align with the requirement and the interpretation thereof.

Reply:
> Yes. There is some juggling going on to get the code to align with the requirement and the interpretation thereof.

Is this stable now, specifically the "Selecting a main method" section and the behavior change where an instance main takes precedence over an "inherited" static main? Just checking before doing another pass over the method finder.

Reply:
I think so - additional changes will be coming from Joe via https://github.com/openjdk/jdk/pull/14165